### PR TITLE
Refactor browsers

### PR DESCRIPTION
**Pull request by niklassiemer**
The `DataContainer` e.g. does not have a `path` to browse through it. The base `HasGroupsBrowser` only assumes the most general features of the `HasGroups` mixin. Therefore, this browser should work for every `PyironObject`! I also used a lighter design with all groups/nodes on one line (with wrapping to another line if too long). 

There are also some additional classes:
- `HasGroupBrowserWithOutput`: Adding an output window to the `HasGroupsBrowser`
- `HasGroupsBrowserWithHistoryPath`: Adding a path derived from the history to the `HasGroupsBrowser`
- `ProjectBrowser`: Using the pyiron internal `path`/`project_path` attributes of all `Project`s to implement a path on top of the `HasGroupsBrowserWithOutput`

@pmrv 
With these changes it should be easy to implement a `DataContainerGUI` which could be based on `HasGroupBrowserWithOutput` and `HasGroupsBrowserWithHistoryPath` and add some functionality to change the clicked data field.

Tests need to be written/corrected!

### CODE REVIEW

**Comment by niklassiemer** [comment]
ping @liamhuber @pmrv

File: pyiron_gui/project/project_browser.py [hunk 1]
```diff
@@ -69,16 +72,19 @@ def display(self, obj, default_output=None):
...
-                raise TypeError("Given 'obj' is of 'NoneType'.")
-            elif self._display_obj is None:
-                print(default_output)
-            elif isinstance(self._display_obj, ObjectWidget):
-                display(self._display_obj.gui)
-            else:
-                plt.ioff()
-                display(self._output_conv())
+        if isinstance(self._display_obj, widgets.DOMWidget):
+            self.box.children = tuple([self._display_obj])
```
Comment (pmrv) [inline pyiron_gui/project/project_browser.py:76]:
```suggestion
            self.box.children = (self._display_ob,)
```

File: pyiron_gui/project/project_browser.py [hunk 2]
```diff
@@ -123,7 +131,363 @@ def _output_conv(self):
...
+            self._history_idx = hist_idx
+        self._project = self._history[self._history_idx]
+        self._update_groups_and_nodes()
+        self.refresh()
+
+    def _go_back(self, _=None):
+        self._history_idx -= 1
+        self._load_history()
+
+    def _go_forward(self, _=None):
```
Comment (pmrv) [inline pyiron_gui/project/project_browser.py:251]: What's those arguments for?

File: pyiron_gui/project/project_browser.py [hunk 3]
```diff
@@ -123,7 +131,363 @@ def _output_conv(self):
...
+        self.refresh()
+        return self._box
+
+    def _ipython_display_(self):
+        """Function used by Ipython to display the object."""
+        display(self.gui())
+
+
+class HasGroupsBrowserWithHistoryPath(HasGroupsBrowser):
+    def __init__(self, project, box=None):
```
Comment (pmrv) [inline pyiron_gui/project/project_browser.py:367]: If I see this correctly the normal `HasGroupsBrowser` already has the infrastructure to keep track of the history, what's the difference to this then?

**Review by pmrv — COMMENTED** [review]
Left a few minor things, but generally looks good.  I'll give it a spin later to see how it looks, though.

File: pyiron_gui/project/project_browser.py (hunk 3 above)
Comment (niklassiemer) [reply pyiron_gui/project/project_browser.py:367]: The `HasGroupsBrowser` just keep it's history, while this one constructs an artificial path from the history and displays it in addition.

File: pyiron_gui/utils/busy_check.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+import functools
+import ipywidgets as widgets
+
+__author__ = "Niklas Siemer"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
```
Comment (liamhuber) [inline pyiron_gui/utils/busy_check.py:9]:
```suggestion
    "Copyright 2021, Max-Planck-Institut für Eisenforschung GmbH - "
```

possibly elsewhere too, I only noticed it here.

**Review by liamhuber — APPROVED** [review]
I'm afraid I had to rush through a little bit. I won't be able to look in more detail until next week, but at a glance I don't see anything that worries me so feel free to merge before then if you don't want to wait. Added a couple minor comments.

File: pyiron_gui/project/project_browser.py [hunk 5]
```diff
@@ -123,7 +131,363 @@ def _output_conv(self):
...
+        new._history = self._history.copy()
+        new._history_idx = self._history_idx
+        return new
+
+    def copy(self):
+        """Copy of the browser using a new Vbox."""
+        return self.__copy__()
+
+    @property
+    def groups(self):
```
[the remark this answers is not in the document]
Comment (niklassiemer) [reply pyiron_gui/project/project_browser.py:196]: I just realized, I would need to implement a `__getitem__` to make it a `HasGroups`. Could, however, forward to the `project.__getitem__`.

File: pyiron_gui/project/project_browser.py (hunk 2 above)
Comment (niklassiemer) [reply pyiron_gui/project/project_browser.py:251]: Nice idea! I will look into it 👍

**Comment by niklassiemer** [comment]
I am happy with the PR as it is now! I will leave it a few days in case you want to have another glance at it and merge it next week.

**Pull request merged**